rfd/0174-kube-rbac-bootstrap.md

```diff
@@ -0,0 +1,346 @@
+automatically provisioned to Kubernetes clusters.
+
+## Why
+
+Currently, users have to rely on third-party tools or manual actions
+when making sure their Kubernetes clusters have the correct roles/role bindings and that those are in sync with the RBAC setup defined in Teleport.
+The changes we propose will make it easier for clients to start using Teleport for accessing Kubernetes clusters.
+
+## Scope
+
+This RFD is focused only on Kubernetes resources related to RBAC functionality, other types of resources are out of scope
+and we will not allow them to be provisioned. Although other resources might be added in the future.
+
+## Details
+
+### Improving Day 1 experience - Adding capability to manage Kubernetes RBAC from Teleport. 
+
+Currently users rely on `kubernetes_groups` and/or `kubernetes_users` fields of the role definition when setting up
+access for Kubernetes. This requires roles/bindings to be created by the client in Kubernetes cluster independently from the 
+RBAC setup in the Teleport. These roles/bindings also should be kept in sync with the Teleport RBAC side,
+which creates additional hurdles for clients, especially the new ones, when using Kubernetes access. 
+To improve this situation we will introduce auto-provisioning of the Kubernetes RBAC based on 
+the Teleport roles, which means users will be able to define required permissions completely in the Teleport, and under the hood
+we will keep them in sync transparently for the users.
+
+We will add a new field to the role definition - `kubernetes_permissions`. That field will contain desired Kubernetes RBAC permissions
+for that role and will be similar in structure to the native Kubernetes RBAC:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: kube-role
+spec:
+  allow:
+    kubernetes_labels:
+      "*": "*"
+    kubernetes_groups: ["kube_group"]
+    kubernetes_permissions:
+      namespaces: ["namespace1", "namespace2"]
+      rules:
+      - resources: ["pod", "pod/exec"]
+        verbs: ["get", "list", "create"]
+      - resources: ["deployments"]
+        apiGroups: ["app"]
+        verbs: ["*"]
+      - resources: ["secrets"]
+        apiGroups: [""]
+        resourceNames: ["secret1", "secret2"]
+        verbs: ["get", "list""]
+```
+
+Presence of this field will signal that user wants Teleport to auto-provision these permissions into the Kubernetes clusters.
+Teleport will translate permissions defined in that field into Kubernetes RBAC definitions and will automatically create roles and bindings
+on the target Kubernetes clusters (as specified by the labels). If `namespaces` field includes '*' Teleport will create ClusterRole/ClusterRoleBinding
+pair on the Kubernetes cluster, otherwise Role/Binding pair will be created in the each listed namespace. 
+
+Subject for the bindings will be a group named `teleport:*RoleName*`, e.g. for a Teleport role `kube-role` the resulting 
```

Comment:
```suggestion
Subject for the bindings will be a group named `teleport:md5(clusterName,roleName)`, e.g. for a Teleport role `kube-role` the resulting 
```

We should hash the role and keep it under 63 chars because kube forces us to

Reply:
Ok, then we can do `teleport:clusterName:roleName` by default and if it exceeds 63 characters we switch to `teleport:hash(clusterName,rolename)`, it would leave most of times role names to be human readable.